lib/immutable-struct.rb

```diff
@@ -36,26 +36,47 @@ class ImmutableStruct
   #     p.minor    # => "yup"
   #     p.minor?   # => true
   #
-  def self.new(*attributes,&block)
+  # === Derived attributes
+  # A derived attribute is a zero-arity method (ie. getter) that applies a transformation to the
+  # existing values stored inside an +ImmutableStruct+. For instance:
+  #
+  #    ImmutableStruct.new(:foo, :bar) do
+  #      def derived; self.foo + ":" + self.bar; end
+  #    end
+  #
+  # Please note that, unless overriden, +ImmutableStruct#to_h+ will serialize all zero-arity methods.
+  #
+  #
+  def self.new(*attributes, &block)
+    raise ArgumentError if attributes && Array(attributes).empty?
+
     klass = Class.new do
+      array_attr = lambda { |attr| attr.kind_of?(Array) && attr.size == 1 }
+      boolean_attr = lambda { |attr| String(attr).match(/(^.*)\?$/) }
+
       attributes.each do |attribute|
-        if attribute.to_s =~ /(^.*)\?$/
+        case attribute
+          when ''
```

Comment:
I guess what I'm saying is that no one should ever expect `ImmutableStruct.new('')` to do anything sensible, so it might be simpler to just ignore this case entirely.